rust/agama-server/src/web/common.rs

```diff
@@ -108,3 +116,120 @@ async fn build_service_status_proxy<'a>(
+/// async fn hello(state: State<HelloWorldState>) {};
+///
+/// #[derive(Clone)]
+/// struct HelloWorldState {};
+///
+/// let dbus = connection().await.unwrap();
+/// let progress_router = progress_router(
+///   &dbus, "org.opensuse.HelloWorld", "/org/opensuse/hello"
+/// ).await.unwrap();
+/// let router: Router<HelloWorldState> = Router::new()
+///   .route("/hello", get(hello))
+///   .merge(progress)
+///   .with_state(HelloWorldState {});
+/// });
+/// ```
+///
+/// * `dbus`: D-Bus connection.
+/// * `destination`: D-Bus service name.
+/// * `path`: D-Bus object path.
+pub async fn progress_router<T>(
+    dbus: &zbus::Connection,
+    destination: &str,
+    path: &str,
+) -> Result<Router<T>, ServiceError> {
+    let proxy = build_progress_proxy(dbus, destination, path).await?;
+    let state = ProgressState { proxy };
+    Ok(Router::new()
+        .route("/progress", get(progress))
+        .with_state(state))
+}
+
+#[derive(Clone)]
+struct ProgressState<'a> {
+    proxy: ProgressProxy<'a>,
+}
+
+async fn progress(State(state): State<ProgressState<'_>>) -> Json<Progress> {
+    let proxy = state.proxy;
+    let progress = Progress::from_proxy(&proxy).await.unwrap();
+    Json(progress)
+}
+
+#[pin_project]
+pub struct ProgressStream<'a> {
+    #[pin]
+    inner: PropertyStream<'a, (u32, String)>,
+    proxy: ProgressProxy<'a>,
+}
+
+pub async fn progress_stream<'a>(
+    dbus: zbus::Connection,
+    destination: &'static str,
+    path: &'static str,
+) -> ProgressStream<'a> {
+    let proxy = build_progress_proxy(&dbus, destination, path)
+        .await
+        .unwrap();
+    ProgressStream::new(proxy).await
+}
+
+impl<'a> ProgressStream<'a> {
+    pub async fn new(proxy: ProgressProxy<'a>) -> Self {
+        let stream = proxy.receive_current_step_changed().await;
+        ProgressStream {
+            inner: stream,
+            proxy,
+        }
+    }
+}
+
+impl<'a> Stream for ProgressStream<'a> {
+    type Item = Event;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let pinned = self.project();
+        match pinned.inner.poll_next(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(_change) => match Progress::from_cached_proxy(&pinned.proxy) {
+                Some(progress) => {
+                    let event = Event::Progress {
+                        progress,
+                        service: pinned.proxy.destination().to_string(),
+                    };
+                    Poll::Ready(Some(event))
+                }
+                _ => Poll::Pending,
+            },
+        }
+    }
+}
+
+async fn build_progress_proxy<'a>(
+    dbus: &zbus::Connection,
+    destination: &str,
+    path: &str,
+) -> Result<ProgressProxy<'a>, zbus::Error> {
+    let proxy = ProgressProxy::builder(&dbus)
+        .destination(destination.to_string())?
+        .path(path.to_string())?
```

Comment:
I am not sure if there is advantage of using `&str` when you inside always needs to create new string. If you have String in method call already, then it can be passed or created outside and just passed.

Reply:
Well, using `&str` in the boundaries looks more flexible to me. And, after all, from outside you do not need to know what is going to happen inside. But yes, I could reconsider it.